### ISSUE TITLE
SPOSet API cannot allow modification of underlying SplineInst

### ISSUE THREAD

**Issue by PDoakORNL** [QMCPACK/qmcpack#3980]
**Describe the bug**
BSplineSet.makeClone() -> SplineR2R Copy Constructor copies SplineInst shared ptr
Then later we call an SPOSet method that modifies SplineInst such as applyRotation.  This actually modifies the state of all SPOSets in a clone group.

So if I have the reasonable looking:
```
for (auto& twf : walkers.getTwfs()) {
  twf.getSPOSet("electrons").applyRotation(some_rotation)
}
```
If all those twfs were made by cloning the golden wavefunction
Then all the SPOSets get rotated by number of clones * some_rotation.

You would need to know explicitly which SPOSets are backed by which SplineInst and call just one for each SplineInst.  This is a terrible break of encapsulation.

This is a significant design issue. I think it stems from either 
1. applyRotation is in the wrong interface.  There needs to be an interface at the "SPOBasis" level and there needs to be 1 handle to that per group of clones.
2.  spline coefficients should be pulled up into the SplineX2X level

This really depends on how the optimization algorithm doing the rotation is intended to work.

**To Reproduce**
Need a reproducer.

**Comment by ye-luo** [comment]
I disagree. This is a designed feature. All the variational parameters should be under shared_ptr.
All the SPOSet objects with the same name must be cloned from the same object and all the variational coefficient should be indistinguishable. Thus once it is optimized, the effect should be propagated to all the objects.

**Comment by ye-luo** [comment]
"Then all the SPOSets get rotated by number of clones * some_rotation." Did you forget the "gold" object, the optimizer is supposed to touch that one instead of bookkeeping individual clones.

**Comment by PDoakORNL** [comment]
I think that the global operations belong in a separate interface.  The SPOSet interface should not be both a walker/crowd multiplicity interface and an interface to a global set of parameters/SPOBasis shared between the whole collection of SPOSet's cloned from the explicitly constructed golden SPOSets in the golden TWF.

In the short term the interface should be strongly and clearly documented (i.e. methods should be grouped) because no it doesn't seem natural to me to modify the golden wavefunction and expect that every object historically cloned from it will be updated because at clone time they copy a shared pointer. I was first introduced to this idea of sharing the spline inst or C for LCAO as a way to save memory not induce side effects globally. The global interface also should be clear that with respect to concurrency they must be no reads by clones during the update of the parameters/SPOBasis, because modifications of individual SPOSets are assumed be thread safe.
